File: y2018/python/day18.py
import numpy as np
import numpy.typing as npt

OPEN = '.'
TREE = '|'
YARD = '#'

P1_TIME = 10
P2_TIME = 10 ** 9


def mat_to_str(mat: npt.NDArray[str]):
    return ''.join(''.join(row) for row in mat)


def resource_value(grid: npt.NDArray[str]):
    return (grid == TREE).sum() * (grid == YARD).sum()


area = []
with open('input/day18.txt') as read:
    for r in read.readlines():
        r = r.strip()
        assert all(c in [OPEN, TREE, YARD] for c in r)
        area.append(list(r))
area = np.array(area)

time = 0
cyc_start = -1
seen = {mat_to_str(area): time}
order = [mat_to_str(area)]
while True:
    init = area.copy()
    for r in range(len(area)):
        for c in range(len(area[r])):
            adj = init[max(r - 1, 0):r + 2, max(c - 1, 0):c + 2]
            open_amt = (adj == OPEN).sum() - (area[r, c] == OPEN)
            tree_amt = (adj == TREE).sum() - (area[r, c] == TREE)
            yard_amt = (adj == YARD).sum() - (area[r, c] == YARD)

            if area[r, c] == OPEN:
                if tree_amt >= 3:
                    area[r, c] = TREE
            elif area[r, c] == TREE:
                if yard_amt >= 3:
                    area[r, c] = YARD
            elif area[r, c] == YARD:
                if not (tree_amt > 0 and yard_amt > 0):
                    area[r, c] = OPEN

    time += 1
    if time == P1_TIME:
        print(f"resource value after {P1_TIME} mins: {resource_value(area)}")

    str_area = mat_to_str(area)
    if str_area in seen:
        cyc_start = seen[str_area]
        break
    order.append(str_area)
    seen[str_area] = time

cyc_len = len(seen) - cyc_start
equal_to = cyc_start + ((P2_TIME - cyc_start) % cyc_len)
final_array = np.array(list(order[equal_to])).reshape(area.shape)
print(f"resource value after {P2_TIME} minutes: {resource_value(final_array)}")

File: y2018/rust/src/bin/day18.rs
use std::collections::HashMap;
use std::fs;

const P1_TIME: usize = 10;
const P2_TIME: usize = 1e9 as usize;

#[derive(Debug, Eq, PartialEq, Hash, Copy, Clone)]
enum Cell {
    Open,
    Trees,
    Yard,
}

fn neighbors(r: i32, c: i32) -> Vec<(i32, i32)> {
    vec![
        (r + 1, c),
        (r - 1, c),
        (r, c + 1),
        (r, c - 1),
        (r + 1, c + 1),
        (r + 1, c - 1),
        (r - 1, c + 1),
        (r - 1, c - 1),
    ]
}

fn resource_value(grid: &Vec<Vec<Cell>>) -> u32 {
    let mut tree_num = 0;
    let mut yard_num = 0;
    for c in grid.iter().flatten() {
        match c {
            Cell::Trees => tree_num += 1,
            Cell::Yard => yard_num += 1,
            _ => ()
        }
    }
    tree_num * yard_num
}

fn main() {
    let mut grid: Vec<Vec<_>> = fs::read_to_string("../input/day18.txt")
        .unwrap()
        .lines()
        .map(|r| {
            r.chars()
                .map(|c| match c {
                    '#' => Cell::Yard,
                    '|' => Cell::Trees,
                    _ => Cell::Open,
                })
                .collect()
        })
        .collect();

    let mut time = 0;
    let cyc_start;
    let mut seen = HashMap::from([(grid.clone(), time)]);
    let mut order = vec![grid.clone()];
    loop {
        let mut new_grid = vec![vec![Cell::Open; grid[0].len()]; grid.len()];
        for r in 0..grid.len() {
            for c in 0..grid[r].len() {
                let mut tree_amt = 0;
                let mut yard_amt = 0;
                for (nr, nc) in neighbors(r as i32, c as i32) {
                    if 0 <= nr
                        && nr < grid.len() as i32
                        && 0 <= nc
                        && nc < grid[nr as usize].len() as i32
                    {
                        match grid[nr as usize][nc as usize] {
                            Cell::Trees => tree_amt += 1,
                            Cell::Yard => yard_amt += 1,
                            _ => ()
                        }
                    }
                }

                new_grid[r][c] = grid[r][c];
                match grid[r][c] {
                    Cell::Open => {
                        if tree_amt >= 3 {
                            new_grid[r][c] = Cell::Trees;
                        }
                    }
                    Cell::Trees => {
                        if yard_amt >= 3 {
                            new_grid[r][c] = Cell::Yard;
                        }
                    }
                    Cell::Yard => {
                        if !(tree_amt > 0 && yard_amt > 0) {
                            new_grid[r][c] = Cell::Open;
                        }
                    }
                }
            }
        }

        time += 1;
        grid = new_grid;

        if time == P1_TIME {
            println!("resource value after {P1_TIME} mins: {}", resource_value(&grid));
        }

        if seen.contains_key(&grid) {
            cyc_start = seen[&grid];
            break;
        }
        order.push(grid.clone());
        seen.insert(grid.clone(), time);
    }

    let cyc_len = seen.len() - cyc_start;
    let equal_to = cyc_start + ((P2_TIME - cyc_start) % cyc_len);
    let p2_result = &order[equal_to];
    println!("resource value after {P2_TIME} mins: {}", resource_value(p2_result));
}
